File: veronica_mars.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
#include <set>
#include <map>
#include <climits>
using namespace std;


int main() {
    /* Enter your code here. Read input from STDIN. Print output to STDOUT */   
    int num = 0;
    cin >> num;
    int numer = pow(10, 9)+7;
    set<int> input; 
    map< pair<int,int>, long long> tree;
    tree[make_pair(INT_MIN, INT_MAX)] = 1;
    
    for(int i=0; i<num; ++i)
    {
        int val;
        cin >> val;
        input.insert(val);
        
        int l = 0;
        int r = 0;
        set<int>::iterator it = input.find(val);
        if (it==input.begin() )
            l = INT_MIN;
        else
        {
            it--;
            l = *it;
            it++;
        }
        it++;
        if (it==input.end())
            r = INT_MAX;
        else
            r = *it;
        
        map< pair<int,int>, long long>::iterator itt = tree.find(make_pair(l,r));
        cout << itt->second;
        if (i==num-1)
            cout << "\n";
        else
            cout << " ";
        
        tree[make_pair(l,val)] = (itt->second*2 ) % numer;
        tree[make_pair(val,r)] = ((itt->second)*2+1) % numer;
        tree.erase(itt);
        
    }

    return 0;
}

File: VeronicaMars.java
mport java.io.*;
import java.util.*;

class Range {
        private Integer left;
        private Integer right;
        public Range(int l, int r){
           left = l;
           right = r;
        }
        public Integer getLeft() {return left; }
        public Integer getRight() { return right; }
        public String toString() {
        	return "Left:" + getLeft() + ",Right:" + getRight();
        }
        
        @Override
        public boolean equals(Object o){
        	if (o==null)
        		return false;
        	if (this.getClass()!=o.getClass())
        		return false;
        	
        	Range e = (Range)o;
        	return e.left.equals(this.left) && e.right.equals(this.right);
        }
        
        @Override
        public int hashCode() {
        	return left.hashCode() + right.hashCode();
        }
    }

public class VeronicaMars {
 
    public static void main(String[] args) {
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution. */
        Scanner scan = new Scanner ( System.in );
        int num = scan.nextInt();
        
        Range start = new Range(Integer.MIN_VALUE, Integer.MAX_VALUE);
        TreeSet<Integer> nums = new TreeSet<Integer> ();
        HashMap<Range, Integer> indextree = new HashMap<Range, Integer> ();
        indextree.put( start, 1);
        int modn = (int)Math.pow(10,9) + 7;
        
        for(int i=0; i<num; ++i){
            int value = scan.nextInt();
            nums.add(value);
            
            int l = 0;
            int r = 0;
            int prev = Integer.MIN_VALUE;
            Iterator<Integer> index = nums.iterator();
            
            boolean sign = false;
            while(index.hasNext() && !sign){
                int temp = 0;
                if ( (temp=index.next())==value ){
                    l = temp == nums.first() ? Integer.MIN_VALUE : prev;
                    r = temp == nums.last() ? Integer.MAX_VALUE : index.next();
                    sign = true;
                }                
                prev = temp;
            }
            
            Range or = new Range(l,r);
            Range lr = new Range(l, value);
            Range rr = new Range(value, r);
            Integer result = indextree.get(or);
            if (i==num-1)
                System.out.println(result);
            else
                System.out.print( result + " ");
            
            indextree.put(lr, (result*2)%modn);
            indextree.put(rr, (result*2+1)%modn);
            indextree.remove(or);
        }
    }
}
